src/extension/infer.rs

```diff
@@ -0,0 +1,855 @@
+    /// graph, either by making a new meta, or looking it up
+    fn make_or_get_meta(&mut self, node: Node, dir: Direction) -> Meta {
+        if let Some(m) = self.extensions.get(&(node, dir)) {
+            *m
+        } else {
+            let m = self.fresh_meta();
+            self.extensions.insert((node, dir), m);
+            m
+        }
+    }
+
+    /// Iterate over the nodes in a hugr and generate unification constraints
+    fn gen_constraints<T>(&mut self, hugr: &T)
+    where
+        T: HugrView,
+    {
+        // The toplevel sibling graph can be open, and we should note what those variables are
+        let toplevel: SiblingGraph<Node, T> = SiblingGraph::new(hugr, hugr.root());
+        for toplevel_node in toplevel.nodes() {
+            let m_input = self.make_or_get_meta(toplevel_node, Direction::Incoming);
+            self.variables.insert(m_input);
+        }
+
+        for node in hugr.nodes() {
+            let m_input = self.make_or_get_meta(node, Direction::Incoming);
+            let m_output = self.make_or_get_meta(node, Direction::Outgoing);
+
+            let node_type = hugr.get_nodetype(node);
+
+            // Add constraints for the inputs and outputs of dataflow nodes according
+            // to the signature of the parent node
+            if let Some([input, output]) = hugr.get_io(node) {
+                for dir in Direction::BOTH {
+                    let m_input_node = self.make_or_get_meta(input, dir);
+                    self.add_constraint(m_input_node, Constraint::Equal(m_input));
+                    let m_output_node = self.make_or_get_meta(output, dir);
+                    self.add_constraint(m_output_node, Constraint::Equal(m_output));
+                }
+            }
+
+            match node_type.signature() {
+                // Input extensions are open
+                None => {
+                    self.gen_union_constraint(
+                        m_input,
+                        m_output,
+                        node_type.op_signature().extension_reqs,
+                    );
+                }
+                // We have a solution for everything!
+                Some(sig) => {
+                    self.add_solution(m_input, sig.input_extensions.clone());
+                    self.add_solution(m_output, sig.output_extensions());
+                }
+            }
+        }
+        // Seperate loop so that we can assume that a metavariable has been
+        // added for every (Node, Direction) in the graph already.
+        for tgt_node in hugr.nodes() {
+            let sig: &OpType = hugr.get_nodetype(tgt_node).into();
+            // Incoming ports with a dataflow edge
+            for port in hugr
+                .node_inputs(tgt_node)
+                .filter(|src_port| matches!(sig.port_kind(*src_port), Some(EdgeKind::Value(_))))
+            {
+                for (src_node, _) in hugr.linked_ports(tgt_node, port) {
```

Comment:
Can this nested for loop be replaced with a call to `hugr.all_neighbours(tgt_node)`

Reply:
No, because we specifically want to find nodes connected by _value edges_, not just neighbours